### DISCUSSION TITLE
Help: How to test changes in an existing integration (fromZigbee, toZigbee)

### DISCUSSION THREAD

**Discussion by mgarea** [Koenkk/zigbee2mqtt#5421] [General]
Hi everyone,

I have purchased a [Lonsonho curtain switch](https://www.zigbee2mqtt.io/devices/11830304.html) and I want to add support for calibration, but I need some help to do the tests.
The [Tuya version of the switch](https://www.zigbee2mqtt.io/devices/TS130F.html) supports calibration. However, the Lonsonho version does not support it through the Zigbee2MQTT addon (It only exposes `cover_position`).

I'm using the `hassio-zigbee2mqtt` add-on for Home Assistant. I know Zigbee2MQTT uses `zigbee-herdsman-converters` to communicate with devices. I read all the documentation and I want to modify an integration by manually editting `devices.js`, `fromZigbee.js` and `toZigbee.js`. However I don't know it that's even possible nor how to do it.

I read many discussions about how to customize the integration, but I don't know how to access files `devices.js`, `fromZigbee.js` and `toZigbee.js` in my Hassio installation.
I found some information about how to add support for new devices and how to define external converters, but I just want to edit an existing integration:

- [Adding support for new devices](https://github.com/danielwelch/hassio-zigbee2mqtt#adding-support-for-new-devices)
- [External converters configuration](https://www.zigbee2mqtt.io/information/configuration.html#external-converters-configuration)
- [How to support new Tuya devices](https://www.zigbee2mqtt.io/how_tos/how_to_support_new_tuya_devices.html)

That said, Is it possible to edit an existing integration by manually modifying `devices.js`, `fromZigbee.js` and `toZigbee.js`? How can I do that?
I'm thinking about using the external converters approach to overwrite the current device. Will that work?

As a side note, I found an interesting discussion about the integration of this device [here](https://github.com/Koenkk/zigbee2mqtt/issues/4257), and I wanted to test what some users propose in that topic.

Thanks in advance.

Best,
Manuel

**Comment by Koenkk** [comment]
The easiest way to do this is by using a bare metal installation (https://www.zigbee2mqtt.io/getting_started/running_zigbee2mqtt.html). Then devices.js is available under /opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/devices.js
